cumulus/primitives/storage-weight-reclaim/src/lib.rs

```diff
@@ -175,10 +175,25 @@ where
 		let storage_size_diff =
 			benchmarked_weight.saturating_sub(unspent).abs_diff(consumed_weight as u64);
 
+		let extrinsic_len = frame_system::AllExtrinsicsLen::<T>::get().unwrap_or(0);
+		let node_side_pov_size = post_dispatch_proof_size.saturating_add(extrinsic_len.into());
+
 		// This value will be reclaimed by [`frame_system::CheckWeight`], so we need to calculate
 		// that in.
 		frame_system::BlockWeight::<T>::mutate(|current| {
-			if consumed_weight > benchmarked_weight {
+			let block_weight_proof_size = current.total().proof_size();
+
+			// If we encounter a situation where the node-side proof size is already higher than
+			// what we have in the runtime bookkeeping, we add the difference to the `BlockWeight`.
+			// This prevents that the proof size grows faster than the runtime proof size.
+			if node_side_pov_size > block_weight_proof_size {
+				let missing = node_side_pov_size.saturating_sub(block_weight_proof_size);
+				current.accrue(Weight::from_parts(0, missing), info.class);
+				log::warn!(
+					target: LOG_TARGET,
+					"Node-side PoV size higher than runtime proof size weight. node-side: {node_side_pov_size} extrinsic_len: {extrinsic_len} runtime: {block_weight_proof_size}, missing: {missing}. Setting to node-side proof size."
+				);
```

Comment:
sounds good indeed.

I also think we could formalize the concept of weight reservation. So that `on_initialize` can reserve some weight for `on_finalize`.
I believe that extrinsic should also be able to reserve some weight if they trigger some logic in `on_finalize`.

Reply:
> I also think we could formalize the concept of weight reservation. So that `on_initialize` can reserve some weight for `on_finalize`.
> I believe that extrinsic should also be able to reserve some weight if they trigger some logic in `on_finalize`.

I don't have a full overview of all the implications of this. But for sure it would help making the math here more accurate. This is one of the reasons why we need to work with the diffs instead of just setting the proof size to the node variant in the first place.